### PR TITLE
dual publish

### LINKED ISSUE #487: Setup dual CJS/ESM packages
So that we can avoid this issue in downstream projects https://github.com/polkadot-js/api/issues/5636

https://nodejs.org/api/packages.html#dual-commonjses-module-packages

### PR DESCRIPTION
**Pull request by qiweiii**
close #487 

build for both cjs and esm.

- [x] test passed
- [x] web-test works
- [x] polkadot.js works

### CODE REVIEW

**Review by xlc — APPROVED** [review]

**Comment by ermalkaleci** [comment]
Only core needs to be dual and this is not dual

**Comment by qiweiii** [comment]
> Only core needs to be dual and this is not dual

@ermalkaleci could you explain a bit

**Comment by ermalkaleci** [comment]
my understanding of dual is when you have `.cjs` and `.mjs` file extension and node will import `.cjs` and if you enable esm it will choose `.mjs` file extension

**Comment by qiweiii** [comment]
> my understanding of dual is when you have `.cjs` and `.mjs` file extension and node will import `.cjs` and if you enable esm it will choose `.mjs` file extension

Yes, this is another way to do this

**Comment by ermalkaleci** [comment]
I don't think we need this anyway. Bundlers will just transform the code and bun for example works with both.

**Comment by xlc** [comment]
> another thing. Do not change current path. I'm sure it will break somewhere. Just add esm into `/lib/esm`. The tests are all running ok because our ts setup doesn't require any pre-build to work

assuming everything are wired correctly (the export map), everything should work

File: packages/chopsticks/tsconfig.esm.json [hunk 1]
```diff
@@ -0,0 +1,8 @@
+{
+	"extends": "./tsconfig.json",
+	"compilerOptions": {
+		"module": "ES2020",
```
Comment (ermalkaleci) [inline packages/chopsticks/tsconfig.esm.json:4]: ES2021 because we use weakref
```suggestion
		"module": "ES2021",
```

**Comment by ermalkaleci** [comment]
keep the current tsconfig and create new one only for esm so vscode won't give warnings

File: packages/chopsticks/tsconfig.esm.json (hunk 1 above)
Comment (qiweiii) [reply packages/chopsticks/tsconfig.esm.json:4]: es2021 is not a valid module option, i will use es2022?

**Comment by qiweiii** [comment]
> keep the current tsconfig and create new one only for esm so vscode won't give warnings

what warnings does it gove?

**Comment by ermalkaleci** [comment]
tools will look for tsconfig.json which doesn't have any reference. we don't need to have 3 tsconfig. Just default one for cjs and another one for esm

**Review by ermalkaleci — APPROVED** [review]

**Pull request merged**